PurchasesCoreSwift/Networking/ETagManager.swift

```diff
@@ -0,0 +1,124 @@
+//
+
+import Foundation
+
+@objc(RCETagManager) public class ETagManager: NSObject {
+    internal static let eTagHeaderName = "X-RevenueCat-ETag"
+
+    private let queue = DispatchQueue(label: "ETagManager")
+
+    private let userDefaults: UserDefaults
+
+    @objc public override init() {
+        self.userDefaults = UserDefaults(suiteName: ETagManager.suiteName) ?? UserDefaults.standard
+    }
+
+    @objc public init(userDefaults: UserDefaults) {
+        self.userDefaults = userDefaults
+    }
+
+    @objc public func eTagHeader(for urlRequest: URLRequest, refreshETag: Bool = false) -> [String: String] {
+        var storedETag = ""
+        if !refreshETag, let storedETagAndResponse = storedETagAndResponse(for: urlRequest) {
+            storedETag = storedETagAndResponse.eTag
+        }
+        return [ETagManager.eTagHeaderName: storedETag]
+    }
+
+    @objc public func getHTTPResultFromCacheOrBackend(with response: HTTPURLResponse,
+                                                      responseObject: [String: Any]?,
+                                                      error: Error?,
+                                                      request: URLRequest,
+                                                      retried: Bool) -> HTTPResponse? {
+        let statusCode = response.statusCode
+        let headersInResponse = response.allHeaderFields
+        let resultFromBackend = HTTPResponse(statusCode: statusCode, responseObject: responseObject)
+        guard error == nil else { return resultFromBackend }
+
+        let eTagInResponse: String? = headersInResponse[ETagManager.eTagHeaderName] as? String ??
+                headersInResponse[ETagManager.eTagHeaderName.lowercased()] as? String
+
+        if eTagInResponse != nil {
+            if shouldUseCachedVersion(responseCode: statusCode) {
+                guard let storedResponse = getStoredHTTPResponse(for: request) else {
+                    if retried {
+                        Logger.warn(
+                                String(
+                                        format: Strings.network.could_not_find_cached_response_in_already_retried,
+                                        resultFromBackend
+                                )
+                        )
+                        return resultFromBackend
+                    } else {
+                        return nil
+                    }
+                }
+                return storedResponse
+            }
+            storeStatusCodeAndResponseIfNoError(
+                    for: request,
+                    statusCode: statusCode,
+                    responseObject: responseObject,
+                    eTag: eTagInResponse!)
+        }
+
+        return resultFromBackend
+    }
+
+    @objc public func clearCaches() {
+        self.userDefaults.removePersistentDomain(forName: ETagManager.suiteName)
+    }
+
+    private func shouldUseCachedVersion(responseCode: Int) -> Bool {
```

Comment:
I usually do `private extension` for these, to have a more obvious separation between public and private stuff, and so that I don't have to explicitly define the privacy per method (everything in a private extension is private). 

It's not a 100% clean win, though, since all the private stored properties will still have to be in the public declaration of the class, so the main declaration ends up having a mixed of public and private anyway